File: ISMemoryCacheTests/ISMemoryCacheTests.m
#import <SenTestingKit/SenTestingKit.h>
#import <OCMock/OCMock.h>
#import "ISMemoryCache.h"

static NSString *ISTestKey = @"test";

@interface ISMemoryCacheTests : SenTestCase {
    ISMemoryCache *cache;
}

@end

@implementation ISMemoryCacheTests

- (void)setUp
{
    [super setUp];
    cache = [[ISMemoryCache alloc] init];
}

- (void)tearDown
{
    cache = nil;
    [super tearDown];
}

- (void)testSetObjectForKey
{
    NSObject *object = [[NSObject alloc] init];
    [cache setObject:object forKey:ISTestKey];
    STAssertEquals([cache objectForKey:ISTestKey], object, @"object for key should equal inserted object.");
    STAssertTrue([[cache allKeys] containsObject:ISTestKey], nil);
}

- (void)testRemoveObjectForKey
{
    NSObject *object = [[NSObject alloc] init];
    [cache setObject:object forKey:ISTestKey];
    [cache removeObjectForKey:ISTestKey];
    STAssertNil([cache objectForKey:ISTestKey], @"cache should not contain object for key.");
    STAssertFalse([[cache allKeys] containsObject:ISTestKey], nil);
}

- (void)testRemoveUnretainedObjects
{
    @autoreleasepool {
        NSObject *unretainedObject = [[NSObject alloc] init];
        [cache setObject:unretainedObject forKey:ISTestKey];
    }
    [cache removeUnretainedObjects];
    STAssertNil([cache objectForKey:ISTestKey], @"object for key should be removed.");
}

- (void)testNotRemoveRetainedObjects
{
    NSObject *retainedObject = [[NSObject alloc] init];
    @autoreleasepool {
        [cache setObject:retainedObject forKey:ISTestKey];
    }
    [cache removeUnretainedObjects];
    STAssertEquals([cache objectForKey:ISTestKey], retainedObject, @"object for key should not be removed.");
}

- (void)testHandleMemoryWarning
{
    id mock = [OCMockObject partialMockForObject:cache];
    [[mock expect] removeUnretainedObjects];
    
    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
    NSString *name = UIApplicationDidReceiveMemoryWarningNotification;
    [center postNotificationName:name object:nil];
}

@end

File: ISMemoryCacheTests/ISMemoryCacheTests.swift
import UIKit
import XCTest
import ISMemoryCache

class ISMemoryCacheTests: XCTestCase {
    var cache: ISMemoryCache!
    
    override func setUp() {
        super.setUp()
        cache = ISMemoryCache()
    }
    
    override func tearDown() {
        cache = nil
        super.tearDown()
    }
    
    func testSetAndGetValue() {
        let key = "key"
        let value = NSObject()
        
        cache[key] = value
        
        if let cachedValue = cache[key] as? NSObject {
            XCTAssertEqual(cachedValue, value)
        } else {
            XCTFail()
        }
    }
    
    func testRemoveValue() {
        let key = "key"
        let value = NSObject()
        
        cache[key] = value
        cache[key] = nil
        
        XCTAssertNil(cache[key])
    }
    
    func testRemoveUnretainedValues() {
        let key = "key"
        
        autoreleasepool {
            self.cache[key] = NSObject()
        }
        
        cache.performClearAction(.RemoveUnretainedObjects)
        
        XCTAssertNil(cache[key])
    }
    
    func testAvoidRemovingRetainedValues() {
        let key = "key"
        let value = NSObject()
        
        autoreleasepool {
            self.cache[key] = value
        }
        
        cache.performClearAction(.RemoveUnretainedObjects)
        
        if let cachedValue = cache[key] as? NSObject {
            XCTAssertEqual(cachedValue, value)
        } else {
            XCTFail()
        }
    }
    
    func testHandleApplicationDidEnterBackgroundNotification() {
        let key = "key"
        
        autoreleasepool {
            self.cache[key] = NSObject()
        }
        
        let notificationCenter = NSNotificationCenter.defaultCenter()
        notificationCenter.postNotificationName(UIApplicationDidEnterBackgroundNotification, object: nil)
        
        XCTAssertNil(cache[key])
    }
    
    func testHandleApplicationDidReceiveMemoryWarningNotification() {
        let key = "key"
        
        autoreleasepool {
            self.cache[key] = NSObject()
        }
        
        let notificationCenter = NSNotificationCenter.defaultCenter()
        notificationCenter.postNotificationName(UIApplicationDidReceiveMemoryWarningNotification, object: nil)
        
        XCTAssertNil(cache[key])
    }
    
    func testThreadSafety() {
        let queue = NSOperationQueue()
        
        for var i = 0; i < 10000; i++ {
            let key = "\(i)"
            let value = NSObject()
            
            queue.addOperationWithBlock {
                self.cache[key] = value
            }
            
            queue.addOperationWithBlock {
                self.cache[key] = self.cache[key]
            }
        }
        
        queue.waitUntilAllOperationsAreFinished()
    }
}
